src/linalg/spectral.jl

```diff
@@ -35,13 +35,13 @@ function adjacency_matrix(g::AbstractGraph, T::DataType=Int; dir::Symbol=:out)
     end
 end
 
-function _adjacency_matrix(g::AbstractGraph{U}, T::DataType, neighborfn::Function, nzmult::Int=1) where U
+function _adjacency_matrix(g::AbstractGraph, T::DataType, neighborfn::Function, nzmult::Int=1)
     n_v = nv(g)
     nz = ne(g) * (is_directed(g) ? 1 : 2) * nzmult
-    colpt = ones(U, n_v + 1)
+    colpt = ones(Int, n_v + 1)
 
-    rowval = sizehint!(Vector{U}(), nz)
-    selfloops = Vector{U}()
+    rowval = sizehint!(Vector{Int}(), nz)
```

Comment:
Why do we need to change `rowval` here?

Reply:
colptr and rowval need to be the same type (i think it is a bug, pls check the comment I made below).

Reply:
As stated here in [documentation](https://docs.julialang.org/en/v1/stdlib/SparseArrays/index.html). 
colptr and rowval are required to be of the same type, Sorry for the confusion, I failed to read the documentation correctly earlier when I thought this was a bug.

Reply:
> 
> 
> > I am not sure though, if it is worth sacrificing low memory usage for type stability ?
> 
> In this case we're O(|V|) worst-case, so I think this is OK.

If rowval is required to be of the same type then I think it will be O(|E||V|) .

Reply:
Would it be possible, to allocate the space for `rowal` immediately and then write to to that array at the correct position? The combination of `sizehint!` and `push! ` is still slower at the moment.

Reply:
@simonschoelly If I understand correctly, are you suggesting -
```
x = 0
for i = 1:nv(g)
    x += length(neighborfn(g, i))
end
rowval = Vector{Int}(undef, x)
```

Reply:
Not sure why you suddently need this `x`, don't you already have `nz` that tells you how big this array should be?
So you could write `rowval = Vector{Int}(undef, nz)`. Or is `nz` not exactly the size of the resulting vector?

Reply:
Okay, I thought nz was an estimate of the size and thus we were using sizehint, but now I think it IS the actual size.